File: typescript/src/ipc-frame.ts
import { canonicalUtf8, type JsonValue, validateJsonResources } from "./canonical.js";
import { IPC_FRAME_HEADER_BYTES, IPC_FRAME_MAX_FEED_CHUNK_BYTES, IPC_FRAME_MAX_PAYLOAD_BYTES, IPC_FRAME_MIN_PAYLOAD_BYTES } from "./generated-runtime-profiles.js";

export const FRAME_HEADER_BYTES = IPC_FRAME_HEADER_BYTES;
export const MIN_FRAME_PAYLOAD_BYTES = IPC_FRAME_MIN_PAYLOAD_BYTES;
export const MAX_FRAME_PAYLOAD_BYTES = IPC_FRAME_MAX_PAYLOAD_BYTES;
export const MAX_FEED_CHUNK_BYTES = IPC_FRAME_MAX_FEED_CHUNK_BYTES;

export class IpcFrameError extends Error {}

export function encodeFrame(value: unknown): Uint8Array {
  const payload = canonicalUtf8(value);
  if (payload.length < MIN_FRAME_PAYLOAD_BYTES || payload.length > MAX_FRAME_PAYLOAD_BYTES) {
    throw new IpcFrameError("canonical payload size is outside A620-IPC-FRAME-1 limits");
  }
  const frame = new Uint8Array(FRAME_HEADER_BYTES + payload.length);
  new DataView(frame.buffer).setUint32(0, payload.length, false);
  frame.set(payload, FRAME_HEADER_BYTES);
  return frame;
}

function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) return false;
  for (let i = 0; i < left.length; i++) if (left[i] !== right[i]) return false;
  return true;
}

export function decodeCanonicalPayload(payload: Uint8Array): JsonValue {
  if (payload.length < MIN_FRAME_PAYLOAD_BYTES || payload.length > MAX_FRAME_PAYLOAD_BYTES) {
    throw new IpcFrameError("frame payload size is outside A620-IPC-FRAME-1 limits");
  }
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(payload);
  } catch (error) {
    throw new IpcFrameError(`frame payload is not valid UTF-8: ${String(error)}`);
  }
  let value: unknown;
  try {
    value = JSON.parse(text);
  } catch (error) {
    throw new IpcFrameError(`frame payload is not valid JSON: ${String(error)}`);
  }
  try {
    validateJsonResources(value);
  } catch (error) {
    throw new IpcFrameError(String(error));
  }
  const canonical = canonicalUtf8(value);
  if (!equalBytes(canonical, payload)) {
    throw new IpcFrameError("frame payload is not A620-JCS-1 canonical JSON");
  }
  return value;
}

/** Streaming decoder retaining only one partial header/payload. */
export class FrameDecoder {
  private readonly header = new Uint8Array(FRAME_HEADER_BYTES);
  private headerBytes = 0;
  private payload: Uint8Array | null = null;
  private payloadBytes = 0;
  private failed = false;

  private fail(message: string): never {
    this.failed = true;
    this.headerBytes = 0;
    this.payload = null;
    this.payloadBytes = 0;
    throw new IpcFrameError(message);
  }

  feed(chunk: Uint8Array): Uint8Array[] {
    if (this.failed) throw new IpcFrameError("frame decoder is already failed");
    if (!(chunk instanceof Uint8Array)) this.fail("frame chunk must be Uint8Array");
    if (chunk.length > MAX_FEED_CHUNK_BYTES) this.fail("frame feed chunk exceeds A620-IPC-FRAME-1 limit");

    const frames: Uint8Array[] = [];
    let offset = 0;
    while (offset < chunk.length) {
      if (this.payload === null) {
        const take = Math.min(FRAME_HEADER_BYTES - this.headerBytes, chunk.length - offset);
        this.header.set(chunk.subarray(offset, offset + take), this.headerBytes);
        this.headerBytes += take;
        offset += take;
        if (this.headerBytes < FRAME_HEADER_BYTES) continue;
        const length = new DataView(this.header.buffer).getUint32(0, false);
        this.headerBytes = 0;
        if (length < MIN_FRAME_PAYLOAD_BYTES || length > MAX_FRAME_PAYLOAD_BYTES) {
          this.fail(`invalid frame payload length: ${length}`);
        }
        this.payload = new Uint8Array(length);
        this.payloadBytes = 0;
      }

      const payload = this.payload;
      const take = Math.min(payload.length - this.payloadBytes, chunk.length - offset);
      payload.set(chunk.subarray(offset, offset + take), this.payloadBytes);
      this.payloadBytes += take;
      offset += take;
      if (this.payloadBytes === payload.length) {
        frames.push(payload);
        this.payload = null;
        this.payloadBytes = 0;
      }
    }
    return frames;
  }

  get retainedBytes(): number {
    return this.headerBytes + this.payloadBytes;
  }

  finish(): void {
    if (this.failed) throw new IpcFrameError("frame decoder is already failed");
    if (this.headerBytes !== 0 || this.payload !== null || this.payloadBytes !== 0) {
      this.fail("channel closed with a trailing partial frame");
    }
  }
}

File: python/a620_gate0/ipc_frame.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .canonical import canonical_bytes, strict_json_loads
from .generated_profiles import (
    IPC_FRAME_HEADER_BYTES, IPC_FRAME_MAX_FEED_CHUNK_BYTES,
    IPC_FRAME_MAX_PAYLOAD_BYTES, IPC_FRAME_MIN_PAYLOAD_BYTES,
)
from .wire import MAX_WIRE_MESSAGE_BYTES, parse_runtime_message

FRAME_HEADER_BYTES = IPC_FRAME_HEADER_BYTES
MIN_FRAME_PAYLOAD_BYTES = IPC_FRAME_MIN_PAYLOAD_BYTES
MAX_FRAME_PAYLOAD_BYTES = IPC_FRAME_MAX_PAYLOAD_BYTES
MAX_FEED_CHUNK_BYTES = IPC_FRAME_MAX_FEED_CHUNK_BYTES
if MAX_FRAME_PAYLOAD_BYTES != MAX_WIRE_MESSAGE_BYTES:
    raise RuntimeError("IPC frame and wire message byte limits diverged")


class IpcFrameError(ValueError):
    pass


def encode_frame(value: Any) -> bytes:
    payload = canonical_bytes(value)
    if not MIN_FRAME_PAYLOAD_BYTES <= len(payload) <= MAX_FRAME_PAYLOAD_BYTES:
        raise IpcFrameError("canonical payload size is outside A620-IPC-FRAME-1 limits")
    return len(payload).to_bytes(FRAME_HEADER_BYTES, "big", signed=False) + payload


def decode_canonical_payload(payload: bytes) -> Any:
    if not MIN_FRAME_PAYLOAD_BYTES <= len(payload) <= MAX_FRAME_PAYLOAD_BYTES:
        raise IpcFrameError("frame payload size is outside A620-IPC-FRAME-1 limits")
    value = strict_json_loads(payload)
    if canonical_bytes(value) != payload:
        raise IpcFrameError("frame payload is not A620-JCS-1 canonical JSON")
    return value


def decode_runtime_payload(payload: bytes) -> dict[str, Any]:
    value = decode_canonical_payload(payload)
    if not isinstance(value, dict):
        raise IpcFrameError("runtime frame root must be an object")
    return parse_runtime_message(payload)


@dataclass
class FrameDecoder:
    """Fail-closed streaming A620-IPC-FRAME-1 decoder.

    The decoder never concatenates an arbitrary read with its retained state.
    It retains at most a partial four-byte header or one declared payload, so a
    read containing several maximum-size frames is processed without a false
    buffer-overflow rejection or an avoidable second full-size copy.
    """

    _header: bytearray = field(default_factory=bytearray)
    _payload: bytearray = field(default_factory=bytearray)
    _expected_payload_bytes: int | None = None
    _failed: bool = False

    def _fail(self, message: str) -> None:
        self._failed = True
        self._header.clear()
        self._payload.clear()
        self._expected_payload_bytes = None
        raise IpcFrameError(message)

    def feed(self, chunk: bytes | bytearray | memoryview) -> list[bytes]:
        if self._failed:
            raise IpcFrameError("frame decoder is already failed")
        if not isinstance(chunk, (bytes, bytearray, memoryview)):
            self._fail("frame chunk must be bytes-like")
        view = memoryview(chunk).cast("B")
        if len(view) > MAX_FEED_CHUNK_BYTES:
            self._fail("frame feed chunk exceeds A620-IPC-FRAME-1 limit")

        frames: list[bytes] = []
        offset = 0
        while offset < len(view):
            if self._expected_payload_bytes is None:
                take = min(FRAME_HEADER_BYTES - len(self._header), len(view) - offset)
                self._header.extend(view[offset : offset + take])
                offset += take
                if len(self._header) < FRAME_HEADER_BYTES:
                    continue
                length = int.from_bytes(self._header, "big", signed=False)
                self._header.clear()
                if not MIN_FRAME_PAYLOAD_BYTES <= length <= MAX_FRAME_PAYLOAD_BYTES:
                    self._fail(f"invalid frame payload length: {length}")
                self._expected_payload_bytes = length

            expected = self._expected_payload_bytes
            assert expected is not None
            take = min(expected - len(self._payload), len(view) - offset)
            self._payload.extend(view[offset : offset + take])
            offset += take
            if len(self._payload) == expected:
                frames.append(bytes(self._payload))
                self._payload.clear()
                self._expected_payload_bytes = None

        return frames

    @property
    def retained_bytes(self) -> int:
        return len(self._header) + len(self._payload)

    def finish(self) -> None:
        if self._failed:
            raise IpcFrameError("frame decoder is already failed")
        if self._expected_payload_bytes is not None or self._header or self._payload:
            self._fail("channel closed with a trailing partial frame")
